src/osd/PrimaryLogPG.cc

```diff
@@ -6700,13 +6700,14 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	        oi.size - op.extent.truncate_size);
 	      ctx->modified_ranges.union_of(trim);
 	      ctx->clean_regions.mark_data_region_dirty(op.extent.truncate_size, oi.size - op.extent.truncate_size);
-	      oi.clear_data_digest();
 	    }
 	    if (op.extent.truncate_size != oi.size) {
               truncate_update_size_and_usage(ctx->delta_stats,
                                              oi,
                                              op.extent.truncate_size);
 	    }
+	    //truncate modify oi.size, need clear old data_digest and DIGEST flag
+	    oi.clear_data_digest();
```

Comment:
This is only necessary if op.extent.truncate_size != oi.size -- let's move this up 2 lines under truncate_update_size_and_usage.

Reply:
...in fact, it should immediately follow every invocation of truncate_update_size_and_usage -- let's just refactor it into truncate_update_size_and_usage.